test/alternator/test_update_expression.py

```diff
@@ -547,6 +551,63 @@ def test_update_expression_list_append(test_table_s):
         ExpressionAttributeNames={'#name1': 'a', '#name2': 'a'})
     assert test_table_s.get_item(Key={'p': p}, ConsistentRead=True)['Item']['a'] == ['dog', 'hi', 2, 4, 'hello', 8]
 
+# A list_append() function requires both its arguments to be lists. If one
+# of them isn't, we get a ValidationException. This also includes the case
+# that one of the arguments is an attribute which doesn't exist, or the
+# entire item doesn't exist.
+# The specific error message returned by Alternator and DynamoDB are not
+# identical - e.g., when an attribute is missing DynamoDB says that ""The
+# provided expression refers to an attribute that does not exist in the item"
+# and Alternator says "list_append() given a non-list" - but the error
+# type, ValidationException, is the same.
+def test_update_expression_list_append_non_list_arguments(test_table_s):
+    p = random_string()
+    # The item p doesn't exist at all, so in particular the attribute "b" is
+    # missing, so in particular b is not a list
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.update_item(Key={'p': p},
+            UpdateExpression='SET a = list_append(b, :val1)',
+            ExpressionAttributeValues={':val1': ['hi', 2]})
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.update_item(Key={'p': p},
+            UpdateExpression='SET a = list_append(:val1, b)',
+            ExpressionAttributeValues={':val1': ['hi', 2]})
+    # The item p does exist, but it doesn't have an attribute b, so
+    # in particular b is not a list:
+    test_table_s.put_item(Item={'p': p, 'a': 2})
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.update_item(Key={'p': p},
+            UpdateExpression='SET a = list_append(b, :val1)',
+            ExpressionAttributeValues={':val1': ['hi', 2]})
+    with pytest.raises(ClientError, match='ValidationException'):
```

Comment:
looks like testing of those ValidationException is similar enough that table test may be shorter and equally (or more) readable. Something like:
```
values = [
"..." // case when ....
"..." // another case ...
]

for v in values:
        test_table_s.put_item(Item={'p': p, 'b': b})
        with pytest.raises(ClientError, match='ValidationException'):
            test_table_s.update_item(Key={'p': p},
                UpdateExpression='SET a = list_append(b, :val1)',
                ExpressionAttributeValues={':val1': ['hi', 2]})
```

Reply:
not sure though if failure message would also be clear (e.g. some way to embellish test error with particular case input, not sure how to do it in python)

Reply:
You're right, it's not pretty that I have exactly the same two statements copied multiple times in this test. It doesn't make this test hard to read - in fact perhaps the opposite (since I have a comment before each repetition explaining what the next repetition tests), but it is quite ugly. I'll think on how to reduce the duplication.

Reply:
I see @denesb already committed this version, so I'll send a followup patch.